### ISSUE TITLE
Can't browse internal memory with home set to SD card

### ISSUE THREAD

**Issue by wssddc-hub** [koreader/koreader#6912]
I set home in koreader to /sdcard/Android/data/org.koreader.launcher on an external SD card formatted FAT32.  That works, but having done this, I see no way to browse to internal memory.

Side notes:
koreader shows this path as /storage/0b17-1ef9/Android/...
Hardware is a Likebook Mars with Android 8.1
I didn't want to tie the SD card to one device. I insert the card in a Windows PC to copy in ebooks.
Discovering that koreader and most other apps could only read files in a directory named for the app was not easy.

**Comment by pazos** [comment]
> Discovering that koreader and most other apps could only read files in a directory named for the app was not easy.

That's usually false. The path `/storage/UUID/Android/data/org.koreader.launcher` is writable and tied to the application (it will be removed if the app is removed). Anything under `/storage/UUID` is readable but not writable.

> I didn't want to tie the SD card to one device. I insert the card in a Windows PC to copy in ebooks.

When used as a external SDCard `/storage/emulated/0` points to the internal partition and `/storage/emulated/UUID` points to the external partition.

The problem is: the parent dir `/storage/emulated` is not readable, so is not possible to switch between the ext sdcard and the internal partition using KO filemanager.

>  Having done this, I see no way to browse to internal memory.

Yup, that's true. A workaround to restore the FM on the internal partition is to open a book stored there (from an external app) and, once opened, press the FM button. This will place the FM on the book path, so you can navigate normally and set the HOME folder to your desired directory in the internal partition.

A better solution would be:

- Use the externalFilesDir array to check the existance of an sdcard:

if it returns **one** path there's no sdcard.
If it returns **two** paths, the second one holds the direction of the sdcard. It would be a matter of removing trailing `Android/data/org.koreader.launcher/files` from that path and we get the full path to the sdcard (again, called `/storage/emulated/UUID`)

With the full path of the external sdcard it is possible to tell the FM to go there. It will still be read only.

PRs welcome. I can help on the android specifics but the FM needs to be changed to show a button/toggle to switch between internal/external partition.

**Comment by pazos** [comment]
AFAICT all the apps that diff between internal partition and sdcard **without** SAF are relying on the same hack, as there're no APIs in android to retrieve the path of the sdcard.

**Comment by pazos** [comment]
@wssddc-hub: could you test that both paths reported in this test apk match what you see on your device. A test with no sdcard inserted is cool too ;)

Attaching here the build APK and its sources -> 
[UserPathsTest.zip](https://github.com/koreader/koreader/files/5604894/UserPathsTest.zip)

It should look like:

![quick](https://user-images.githubusercontent.com/975883/100389955-2e1a3e00-302f-11eb-8899-3d6e91495fbb.png)

**Comment by wssddc-hub** [comment]
Running UserPathsTest with SD card inserted

Internal partition: /storage/emulated/0
External partition: /storage/0B1F-1EF9

With SD card ejected
App stops repeatedly

The results with the SD card inserted match what I see in koreader.

**Comment by wssddc-hub** [comment]
> Yup, that's true. A workaround to restore the FM on the internal partition is to open a book stored there (from an external app) and, once opened, press the FM button. This will place the FM on the book path, so you can navigate normally and set the HOME folder to your desired directory in the internal partition.

This works to reset the HOME folder (using CX Explorer to open a book).  Thanks.

This is a new device, and the SD card is even newer, so I'm experimenting with the best way to configure.  I think I want HOME to be on the SD card.  Navigating from the internal to external locations is a lot of steps.  I tried setting folder shortcuts, but doing so exits koreader and leaves it in a state where I have to kill the task before it will run again.  Perhaps that should be another thread.

**Comment by wssddc-hub** [comment]
> That's usually false. The path `/storage/emulated/UUID/Android/data/org.koreader.launcher` is writable and tied to the application (it will be removed if the app is removed). Anything under `/storage/emulated/UUID` is readable but not writable.

The funny/confusing thing was that from a read-only directory, opening an epub file showed just two letters (PK, the usual start of a zipped file) but a djvu file worked.  (The stock reader software doesn't support djvu, so +1 for koreader although I don't have many files using this format.)

**Comment by pazos** [comment]
> The funny/confusing thing was that from a read-only directory, opening an epub file showed just two letters (PK, the usual start of a zipped file) but a djvu file worked.

I have nothing to comment here. Seems totally unrelated to the issue at hand. If you're able to reproduce **only** while opening the file from an external app and **not** from the FM then please open a different ticket and attach logs there (see below)  :+1: 

>  I tried setting folder shortcuts, but doing so exits koreader and leaves it in a state where I have to kill the task before it will run again.  Perhaps that should be another thread.

I'm interested in this bug. Please open a separate ticket and attach logs. It looks like a native crash, so we need logs catched with: `adb logcat KOReader:I ActivityManager:* AndroidRuntime:* DEBUG:* *:F`

> With SD card ejected App stops repeatedly

Thanks for reporting (and testing): this should fix the issue: 
[app-debug.zip](https://github.com/koreader/koreader/files/5609149/app-debug.zip)

If everything works as expected you'll see "Not available" after extracting the sdcard.


> I think I want HOME to be on the SD card

That's your choice but support will be next to 0 in corner cases (ie: you start KOReader with HOME in sdcard but sdcard is not available or the UUID changed because you inserted a different sdcard)

Also, keep in mind, books opened from RO filesystems will store progress, annotations, etc in a centralized place. If you rename your files you won't be able to retrieve that info. It is recoverable by manually editing these central files and changing paths.

tldr: android is a mess for external sdcards. KOReader will never  work well with them. The **File API** isn't granted to work there and KO relies on that for all platforms it supports. There're no workarounds for that other than writting a new app based on SAF or cripple KO, so it becomes a reader only instead of a full fledged app with its own FM.

**Comment by wssddc-hub** [comment]
> > The funny/confusing thing was that from a read-only directory, opening an epub file showed just two letters (PK, the usual start of a zipped file) but a djvu file worked.
> 
> I have nothing to comment here. Seems totally unrelated to the issue at hand. If you're able to reproduce **only** while opening the file from an external app and **not** from the FM then please open a different ticket and attach logs there (see below) 👍

This is just a quirk, not worth paying attention to.

> > I tried setting folder shortcuts, but doing so exits koreader and leaves it in a state where I have to kill the task before it will run again.  Perhaps that should be another thread.
> 
> I'm interested in this bug. Please open a separate ticket and attach logs. It looks like a native crash, so we need logs catched with: `adb logcat KOReader:I ActivityManager:* AndroidRuntime:* DEBUG:* *:F`

I'll see what I can do.  I do have adb installed, but haven't used it much, and not at all with this device.

> > With SD card ejected App stops repeatedly
> 
> Thanks for reporting (and testing): this should fix the issue:
> [app-debug.zip](https://github.com/koreader/koreader/files/5609149/app-debug.zip)
> 
> If everything works as expected you'll see "Not available" after extracting the sdcard.

Confirmed, it now does the right thing when the SD card is ejected.

> > I think I want HOME to be on the SD card
> 
> That's your choice but support will be next to 0 in corner cases (ie: you start KOReader with HOME in sdcard but sdcard is not available or the UUID changed because you inserted a different sdcard)

I understand the limitations.  It's a 64 GB card, so I'll never overflow it with ebooks and need another.  If the card isn't inserted, the books aren't available so there will be no reason to run any reader software.

**Comment by wssddc-hub** [comment]
> I'm interested in this bug. Please open a separate ticket and attach logs. It looks like a native crash, so we need logs catched with: `adb logcat KOReader:I ActivityManager:* AndroidRuntime:* DEBUG:* *:F`

Crash was a false alarm.  The version I got from F-Droid is old, as in years old.  Or maybe I got it from Boyue's app store, I forget the history, but I've owned the Likebook for less than 2 months.  I just installed 2020.11 from https://github.com/koreader/koreader/releases and folder shortcuts work.

Thanks for your help, and thanks to all developers of this software.  I saw it mentioned favorably multiple times when reading about the Likebook.

**Comment by Frenzie** [comment]
F-Droid offers the most recent release with a delay of a day or two, btw.
So probably not from there.

**Comment by wssddc-hub** [comment]
> F-Droid offers the most recent release with a delay of a day or two, btw. So probably not from there.

It was Boyue's app store. I uninstalled 2020.11 and then installed from Boyue.  That gave me version v2015.11-1868-gb8d9598_2018-10-06.  I'll try to get word back to Boyue.  (I remembered to eject the SD card before doing this test so the org.koreader.launcher directory on the card didn't get removed by the uninstall.)
